positron/doc/positron.1
.\" Process this file with
.\" groff -man -Tascii positron.1
.\" 
.\" 
.TH "positron" "1" "1.0" "" "Xiph.org"
.SH "NAME"
positron \- Neuros Synchronization Manager for Linux

.SH "SYNOPSIS"
.B positron
[
.B add
.I file(s)
]
[
.B sync
]
[
.B sync \-\-reset\-deleted
]
[
.B sync \-\-reset\-recordings
]
[
.B del
.I file(s)
]
[
.B config
]
[
.B rebuild
]
[
.B list
]
[
.B clear
]
[
.B pack
]
[
.B help
.I command
]
[
.B \-m
.I mountpoint
]
[
.B \-c
.I directory
]
.SH "DESCRIPTION"
.B positron
is a synchronization tool for the Neuros portable music player.
.SH "COMMAND SUMMARY"
.IP "add"
Adds files to the Neuros databases, copying as necessary
.IP "sync"
Synchronizes your Neuros with your music library
.IP "del"
Removes files from the database and disk
.IP "config"
Create or modify the current configuration
.IP "rebuild"
Recreates the positron databases
.IP "list"
Lists files in the database
.IP "clear"
Clears all entries from databases without removing any files
.IP "pack"
Removes unused space from databases
.IP "\-m"
Specify a mountpoint
.IP "\-c"
Specify an alternate config directory
.SH "COMMAND DETAIL"
.PP 
.B  positron sync
.RS
Copies any new music tracks from the directories specified
in the configuration file to the Neuros. It also copies new
FM or microphone recordings from the Neuros to your computer.
Positron remembers which tracks have been removed with
.B  positron del
and will not add them again when you sync.
.RE
.PP 
.B  positron sync \-\-reset\-deleted
.RS
Clears the list of deleted files.
.RE
.PP 
.B  positron sync \-\-reset\-recordings
.RS
Clears the list of already\-transferred recordings.
.RE
.PP 
.B  positron add <sourcefile> <targetfile>
.RS
Copies sourcefile to targetfile (path on the Neuros that does not exist).
.RE
.PP 
.B  positron add <sourcefile1> <sourcefile2> ... <targetdir>
.RS
Copies sourcefiles to targetdir (a directory on the Neuros that already exists).
.RE
.PP 
.B  positron add <sourcefile1> <sourcefile2> ...
.RS
Copies sourcefiles to the default music directory.
.RE
.PP 
Files that are already on the Neuros will not be overwritten, but just added to the database if not already present. Non\-music files are ignored.
.PP 
.B  positron del [files or directories]
.RS
Deletes a list of files and/or directories on the Neuros itself (ex: /mnt/neuros/music/rock) to be removed. Both the file on disk and the database entries are deleted.
.RE
.PP 
Only files referenced in the database will be removed, as well as empty directories.
.PP 
.B  positron config
.RS
Asks several questions and creates a new configuration set. Also allows the current configuration to be edited.
.RE
.PP 
.B  positron rebuild
.RS
Rebuilds all the databases from scratch. Positron will locate all of the music files already on the Neuros and put their entries back into your databse as best as it can guess. The database of music stored on your computer ("pcaudio") is cleared and all HiSi clips will be marked as unidentified.
.RE
.PP 
.B  positron list
.RS
List files in the audio database.
.RE
.PP 
.B  positron list <database1> <database2> ...
.RS
Lists all entries from a particular database. Valid database names are: audio, pcaudio, unidedhisi, idedhisi, failedhisi.
.RE
.PP 
.B  positron clear
.RS
Clears all databases.
.RE
.PP 
.B  positron clear <database1> <database2> ...
.RS
Clears all entries from a particular database. Valid database names are: audio, pcaudio, unidedhisi, idedhisi, failedhisi.
.RE
.PP 
The clear command never removes files.
.SH "AUTHORS"
.TP 
Program Author:
.br 
Stan Seibert <volsung@xiph.org>

.TP 
Manpage Author:
.br 
Manuel Lora <vanguardist@xiph.org>
.SH "SEE ALSO"
Move along, nothing to see here. \- Officer Barbrady


positron/ChangeLog
Version 1.0b1 (May 22, 2003):
	* Initial public release

Version 1.0b2 (May 28, 2003):
	* User's Guide and Developer's Guide added
	* Fixed crashing while using "positron del"
	* Albums, Artists, and Genres are now sorted on the Neuros
	* Fixed a problem that caused 35 second MP3s to crash positron
	* Improved MP3 detection
	* Files with illegal characters in the filename or in the tags are
	  handled now.  The characters are changed (usually to underscores)
	  to prevent Neuros lockups.
	* Files are now correctly uploaded to both the Flash and the HD
	  Neuros configurations.  (Before only the HD would work.)


trunk/positron/src/positron.py
#!/usr/bin/env python
import sys
from os import path
import getopt
from config import Config
from neuros import Neuros
import neuros
import cmd_add
import cmd_del
import cmd_list
import cmd_clear
import cmd_pack

import ports

version = "Xiph.org Positron version 0.1"

# Hash table of commands.  The first value in the tuple is the module
# where the command is stored.  The second value is the order to
# display commands in from usage()
commands = { "add"  : (cmd_add,   1),
             "del"  : (cmd_del,   2),
             "list" : (cmd_list,  3),
             "clear": (cmd_clear, 4),
             "pack" : (cmd_pack,  5) }

# For sorting according to the display order element in the tuple
def cmp_func(a, b):
    return a[1] - b[1]

def usage():
    print version, "- Neuros portable music player sync tool"
    print
    
    cmds = commands.values()
    cmds.sort(cmp_func)

    for (item, order) in cmds:
        first_line = item.__doc__.split("\n")[0]
        print "  ",first_line

    print
    print "For more help on a specific command, type: positron help <command>"


def main(argv):
    options = "c:hm:v"
    long_options = ("config=", "help", "mount-point=", "version")

    # parse global options
    try:
        opts, remaining = getopt.getopt(argv[1:], options, long_options)
    except getopt.GetoptError, e:
        print "Error:", e
        usage()
        sys.exit()

    config = Config()
    
    for o,a in opts:
        if o in ("-v", "--version"):
            print version
            sys.exit()
        elif o in ("-h", "--help"):
            usage()
            sys.exit()
        elif o in ("-c", "--config"):
            config.config_filename = a
        elif o in ("-m", "--mount-point"):
            mountpoint = a

    if len(remaining) == 0:
        usage()
        sys.exit(0)
    elif remaining[0] == "help":
        if len(remaining) > 1:
            if commands.has_key(remaining[1]):
                (cmd, display_order) = commands[remaining[1]]
                print cmd.__doc__
            else:
                    print remaining[1], "is not a valid command."
            sys.exit(0)
        else:
            usage()
            sys.exit(0)

    # Read configuration
    try:
        config.read_config_file()
    except IOError,e:
        print "Error opening configuration file \"%s\": %s" \
              % (config.config_file, e)
        print "You can create or modify the configuration file with \"positron config\""
        sys.exit(1)

    # Sanity check
    if not config.mountpoint:
        print "Error: Neuros mountpoint not set with -m and not present in config file."
        print "Please use \"positron config\" to create or modify the configuration."
        sys.exit(1)

    try:
        myNeuros = Neuros(config.mountpoint)

        if commands.has_key(remaining[0]):
            (cmd, display_order) = commands[remaining[0]]
            cmd.run(config, myNeuros, remaining[1:])
        else:
            print remaining[0], "is not a valid command."
            print
            usage()

        exit_value = 0
    except neuros.Error, e:
        print "Error:", e
        exit_value = 1

    sys.exit(exit_value)

    
if __name__ == "__main__":
    main(sys.argv)
